lib/Frontend/ParseableInterfaceSupport.cpp

```diff
@@ -0,0 +1,332 @@
+
+static bool
+extractSwiftInterfaceVersionAndArgs(DiagnosticEngine &Diags,
+                                    clang::vfs::FileSystem &FS,
+                                    StringRef SwiftInterfacePathIn,
+                                    swift::version::Version &Vers,
+                                    llvm::StringSaver &SubArgSaver,
+                                    SmallVectorImpl<const char *> &SubArgs) {
+  auto FileOrError = swift::vfs::getFileOrSTDIN(FS, SwiftInterfacePathIn);
+  if (!FileOrError) {
+    Diags.diagnose(SourceLoc(), diag::error_open_input_file,
+                   SwiftInterfacePathIn, FileOrError.getError().message());
+    return true;
+  }
+  auto SB = FileOrError.get()->getBuffer();
+  auto VersRe = getSwiftInterfaceToolsVersionRegex();
+  auto FlagRe = getSwiftInterfaceModuleFlagsRegex();
+  SmallVector<StringRef, 1> VersMatches, FlagMatches;
+  if (!VersRe.match(SB, &VersMatches)) {
+    Diags.diagnose(SourceLoc(),
+                   diag::error_extracting_version_from_parseable_interface);
+    return true;
+  }
+  if (!FlagRe.match(SB, &FlagMatches)) {
+    Diags.diagnose(SourceLoc(),
+                   diag::error_extracting_flags_from_parseable_interface);
+    return true;
+  }
+  assert(VersMatches.size() == 2);
+  assert(FlagMatches.size() == 2);
+  Vers = swift::version::Version(VersMatches[1], SourceLoc(), &Diags);
+  llvm::cl::TokenizeGNUCommandLine(FlagMatches[1], SubArgSaver, SubArgs);
+  return false;
+}
+
+void
+ParseableInterfaceModuleLoader::configureSubInvocationAndOutputPath(
+    CompilerInvocation &SubInvocation,
+    StringRef InPath,
+    llvm::SmallString<128> &OutPath) {
+
+  auto &SearchPathOpts = Ctx.SearchPathOpts;
+  auto &LangOpts = Ctx.LangOpts;
+
+  // Start with a SubInvocation that copies various state from our
+  // invoking ASTContext.
+  SubInvocation.setImportSearchPaths(SearchPathOpts.ImportSearchPaths);
+  SubInvocation.setFrameworkSearchPaths(SearchPathOpts.FrameworkSearchPaths);
+  SubInvocation.setSDKPath(SearchPathOpts.SDKPath);
+  SubInvocation.setInputKind(InputFileKind::SwiftModuleInterface);
+  SubInvocation.setRuntimeResourcePath(SearchPathOpts.RuntimeResourcePath);
+  SubInvocation.setTargetTriple(LangOpts.Target);
+
+  // Calculate an output filename based on the SubInvocation hash, and
+  // wire up the SubInvocation's InputsAndOutputs to contain both
+  // input and output filenames.
+  OutPath = CacheDir;
+  llvm::sys::path::append(OutPath, llvm::sys::path::stem(InPath));
+  OutPath.append("-");
+  OutPath.append(SubInvocation.getPCHHash());
+  OutPath.append(".");
+  auto Ext = file_types::getExtension(file_types::TY_SwiftModuleFile);
+  OutPath.append(Ext);
+
+  auto &FEOpts = SubInvocation.getFrontendOptions();
+  FEOpts.RequestedAction = FrontendOptions::ActionType::EmitModuleOnly;
+  FEOpts.InputsAndOutputs.addPrimaryInputFile(InPath);
+  SupplementaryOutputPaths SOPs;
+  SOPs.ModuleOutputPath = OutPath.str();
+  StringRef MainOut = "/dev/null";
```

Comment:
I think we need to pick something else here; at the very least this wouldn't be correct on Windows.